### PR TITLE
Network: When using MicroOVN get Northbound DB addresses from `ovn.env`

### LINKED ISSUE #16856: When using MicroOVN LXD should get candidate DB addresses from ovn.env
### Please confirm

- [x] I have searched existing issues to check if an issue already exists for the bug I encountered.

### Distribution

Ubuntu

### Distribution version

24.04

### Output of "snap list --all lxd core20 core22 core24 snapd"

```shell
N/A
```

### Output of "lxc info" or system info if it fails

```shell
N/A
```

### Issue description

MicroCloud configures LXD's https://documentation.ubuntu.com/lxd/latest/server/#server-miscellaneous:network.ovn.northbound_connection at cluster setup time:

@roosterfish 

> MicroCloud queries MicroOVNs `/1.0/services`, filters them for `central` [role] and appends the respective addr to the list of addresses separated by comma

@mkalcok confirms that this a problem if during microovn cluster growth & reorganisation the dynamically assigned `central` role no longer belongs to any of the original members:

> it becomes a problem the moment when DB leader doesn't reside on any of the listed IPs

@mkalcok suggests the correct way for LXD to ascertain the OVN northdb candidate IPs is via the `ovn.env` file provided and maintained by `microovn`:

> microovn exposes ovn.env file via the snap's content interface. That file holds up-to-date values. It's what we use in the microovn internally to run ovn-* commands. That would be my suggestion off the top of my head.

LXD should use its existing detection of MicroOVN (`shared.IsMicroOVNUsed()`) and default to use the `ovn.env` rather than requiring MicroCloud to configure `network.ovn.northbound_connection` which can go stale.

### Steps to reproduce

N/A

### Information to attach

- [ ] Any relevant kernel output (`dmesg`)
- [ ] Instance log (`lxc info NAME --show-log`)
- [ ] Instance configuration (`lxc config show NAME --expanded`)
- [ ] Main daemon log (at `/var/log/lxd/lxd.log` or `/var/snap/lxd/common/lxd/logs/lxd.log`)
- [ ] Output of the client with `--debug`
- [ ] Output of the daemon with `--debug` (or use `lxc monitor` while reproducing the issue)

### PR DESCRIPTION
**Pull request by nmezhenskyi**
This PR updates the `NewOVN()` function to determine the connection string for the Northbound DB based on the environment, if it was not specified by the caller.

If `nbConnection` parameter is empty and MicroOVN snap is used, the `NewOVN()` function will read the `ovn.env` file provided by MicroOVN snap slot [`ovn-env`](https://github.com/canonical/microovn/blob/e860d3abf65c077a1b6524e11459e90e1c10a106/snap/snapcraft.yaml#L22-L26) to determine the connection string.

If `nbConnection` parameter is empty and MicroOVN is NOT used, the `NewOVN()` function will use the `unix:/var/run/ovn/ovnnb_db.sock` as the connection string.

This also updates the default value and description for the `network.ovn.northbound_connection` config option.

Closes https://github.com/canonical/lxd/issues/16856.

This should be merged together with https://github.com/canonical/lxd-pkg-snap/pull/1112.

### CODE REVIEW

File: shared/util_test.go [hunk 1]
```diff
@@ -7,6 +7,7 @@ import (
 	"io"
 	"os"
 	"path/filepath"
+	"reflect"
```
[the remark this answers is not in the document]
Comment (nmezhenskyi) [reply shared/util_test.go]: Fixed

File: shared/util.go [hunk 2]
```diff
@@ -261,6 +261,49 @@ func LogPath(path ...string) string {
...
+		// Split into key and value (only on the first "=").
+		parts := strings.SplitN(line, "=", 2)
+		if len(parts) != 2 {
+			continue
+		}
+
+		key := strings.TrimSpace(parts[0])
+		value := strings.TrimSpace(parts[1])
+
+		// Remove  surrounding quotes if they exist.
```
[the remark this answers is not in the document]
Comment (nmezhenskyi) [reply shared/util.go]: Fixed

**Comment by nmezhenskyi** [comment]
~~I'm not sure I like this current approach because if MicroOVN is used, `NewOVN()` function simply ignores the `nbConnection` parameter and overrides it. It would be more explicit to pass in the correct connection string to the `NewOVN()` when calling it, instead of resolving it inside the function.~~

~~Perhaps I could move the `ovn.env` resolving logic inside the `GlobalConfig.NetworkOVNNorthboundConnection()`? But this doesn't feel quite right as well, since that function should simply return the config value of `network.ovn.northbound_connection`.~~

File: lxd/network/openvswitch/ovn.go [hunk 3]
```diff
@@ -188,6 +192,28 @@ func NewOVN(nbConnection string, sslSettings func() (sslCACert string, sslClient
...
+		ovnEnvVars, err := shared.ParseEnvFile(microOVNEnvFilePath)
+		if err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				return nil, fmt.Errorf("Failed parsing MicroOVN env file %s: %w", microOVNEnvFilePath, err)
+			}
+
+			// Log a warning and fallback to using nbConnection if the file doesn't exist.
+			logger.Warn("MicroOVN is used but ovn.env file is missing, falling back to network.ovn.northbound_connection", logger.Ctx{"err": err})
+		} else {
+			nbConn, ok := ovnEnvVars["OVN_NB_CONNECT"]
```
Comment (simondeziel) [inline lxd/network/openvswitch/ovn.go:219]: Looking at the path used, it seems the LXD doesn't even need to be connected (in the "snap connection" sense) to the MicroOVN snap. If LXD is a snap, shouldn't it use the [`ovn-env` slot](https://github.com/canonical/microovn/blob/e860d3abf65c077a1b6524e11459e90e1c10a106/snap/snapcraft.yaml#L22-L26)?

I think having an environment variable defined should take precedence over opportunistically finding a file and using it.

File: shared/util.go [hunk 4]
```diff
@@ -261,6 +261,49 @@ func LogPath(path ...string) string {
...
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		// Skip empty lines and lines starting with #.
+		if line == "" || strings.HasPrefix(line, "#") {
+			continue
+		}
+
```
Comment (simondeziel) [inline shared/util.go]: At this point I'd probably do a `strings.TrimPrefix(line, "export ")` just to remain compatible with a shell "source-able" file, in case they change their template in the future.

File: shared/util.go [hunk 5]
```diff
@@ -261,6 +261,49 @@ func LogPath(path ...string) string {
...
+		// Split into key and value (only on the first "=").
+		parts := strings.SplitN(line, "=", 2)
+		if len(parts) != 2 {
+			continue
+		}
```
Comment (simondeziel) [inline shared/util.go]: That's probably best done with `strings.Cut(line, "=")`.

File: shared/util_test.go [hunk 6]
```diff
@@ -42,6 +43,91 @@ func TestUrlsJoin(t *testing.T) {
...
+`,
+			expected: map[string]string{"KEY1": "VALUE1", "KEY2": "VALUE2"},
+			wantErr:  false,
+		},
+		{
+			name: "With quotes and comments",
+			content: `
+# This is a comment
+OVN_NB_CONNECT="ssl:10.251.3.75:6641"
+OVN_SB_CONNECT='ssl:10.251.3.75:6642'
```
Comment (simondeziel) [inline shared/util_test.go]:
```suggestion
OVN_SB_CONNECT='ssl:10.251.3.75:6642'  # inline comment
```

File: shared/util_test.go [hunk 7]
```diff
@@ -42,6 +43,91 @@ func TestUrlsJoin(t *testing.T) {
...
+			wantErr: false,
+		},
+		{
+			name:    "Values containing equals signs",
+			content: "KEY=url=http://localhost:8080",
+			expected: map[string]string{
+				"KEY": "url=http://localhost:8080",
+			},
+			wantErr: false,
+		},
```
Comment (simondeziel) [inline shared/util_test.go]: Please add a new one with `export FOO=bar`

File: shared/util.go [hunk 8]
```diff
@@ -261,6 +261,49 @@ func LogPath(path ...string) string {
...
+	}
+
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		// Skip empty lines and lines starting with #.
+		if line == "" || strings.HasPrefix(line, "#") {
```
Comment (simondeziel) [inline shared/util.go]: Since the `#` can technically appear anywhere, I think it'd be safer to use `strings.Cut(line, "#")` and work with what was before the `#`.

File: shared/util_test.go [hunk 9]
```diff
@@ -42,6 +43,91 @@ func TestUrlsJoin(t *testing.T) {
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Create a temporary file.
+			tmpFile, err := os.CreateTemp("", "testenv_")
+			if err != nil {
+				t.Fatalf("Failed creating temp file: %v", err)
+			}
+
+			defer func() { _ = tmpFile.Close() }()
+			defer func() { _ = os.Remove(tmpFile.Name()) }()
```
Comment (simondeziel) [inline shared/util_test.go]: This could be simplified by using a `tmpDir := t.TempDir()` that's automatically cleaned up at the end of the test.

File: lxd/network/openvswitch/ovn.go (hunk 3 above)
Comment (nmezhenskyi) [reply lxd/network/openvswitch/ovn.go:219]: Oh, you're right. How can I use this slot? I assume simply referring to `$SNAP_COMMON` wouldn't work because inside the LXD snap it points to `/var/snap/lxd/common`.

Do I need to add a plug to the LXD snap corresponding to the MicroOVN slot?

File: shared/util.go (hunk 4 above)
Comment (nmezhenskyi) [reply shared/util.go]: Updated

File: shared/util.go (hunk 5 above)
Comment (nmezhenskyi) [reply shared/util.go]: Updated

File: shared/util_test.go (hunk 6 above)
Comment (nmezhenskyi) [reply shared/util_test.go]: Updated

File: shared/util_test.go (hunk 7 above)
Comment (nmezhenskyi) [reply shared/util_test.go]: Updated

File: shared/util.go (hunk 8 above)
Comment (nmezhenskyi) [reply shared/util.go]: Updated

File: shared/util_test.go (hunk 9 above)
Comment (nmezhenskyi) [reply shared/util_test.go]: Updated

File: lxd/network/openvswitch/ovn.go (hunk 3 above)
Comment (simondeziel) [reply lxd/network/openvswitch/ovn.go:219]: I think so but I've not looked into this much. In the `lxd-pkg-snap` repo, you can search for `ovn-certificates` for inspiration.

I'd be happy to explore this with you if needed.

Comment (simondeziel) [reply lxd/network/openvswitch/ovn.go:219]: Even with the (eventual) snap connection, I think the environment variable should take precedence but I'll defer to @tomponline who filled the initial bug report.

Comment (nmezhenskyi) [reply lxd/network/openvswitch/ovn.go:219]: I've updated the approach after the discussion with @tomponline. Now I'm using `$SNAP_DATA` variable to resolve the path to the `ovn.env` file, and I'm adding a new plug to LXD snap for connecting with the `ovn-env` slot provided by MicroOVN.

**Review by simondeziel — COMMENTED** [review]
Lovely, thanks!

**Review by simondeziel — DISMISSED** [review]
Lovely, thanks!

File: lxd/network/openvswitch/ovn.go [hunk 10]
```diff
@@ -180,14 +181,38 @@ type OVNRouterPeering struct {
...
+// If the nbConnection string is empty and MicroOVN is used, the function will read the `ovn.env` file provided by MicroOVN via the snap content interface
+// to determine the connection string for the Northbound DB.
 func NewOVN(nbConnection string, sslSettings func() (sslCACert string, sslClientCert string, sslClientKey string)) (*OVN, error) {
 	// Get database connection strings.
 	sbConnection, err := NewOVS().OVNSouthboundDBRemoteAddress()
 	if err != nil {
 		return nil, fmt.Errorf("Failed getting OVN southbound connection string: %w", err)
 	}
 
+	if nbConnection == "" && shared.IsMicroOVNUsed() {
```
Comment (nmezhenskyi) [inline lxd/network/openvswitch/ovn.go]: The issue with this check here is that `network.ovn.northbound_connection` config option has the default value of `unix:/var/run/ovn/ovnnb_db.sock` .

Since we're always passing the value of  `network.ovn.northbound_connection` inside the `NewOVN()` function, the `nbConnection` parameter is never empty.

Even though the `network.ovn.northbound_connection` option may never be set and does not show up in the `lxc config show`, it is still treated as `unix:/var/run/ovn/ovnnb_db.sock`. Similarly, running `lxc config set network.ovn.northbound_connection=""` does not help.

@tomponline Should we change the behaviour of `network.ovn.northbound_connection` option to have the initial set value of  `unix:/var/run/ovn/ovnnb_db.sock`, but no default one? In that case, clearing the `network.ovn.northbound_connection` config value will result into an empty string.

Comment (tomponline) [reply lxd/network/openvswitch/ovn.go]: @nmezhenskyi Is this true, from what I can see the default value is coming from `getNorthboundDB` which is conditional on whether the `o.nbDBAddr` value is empty?

Is the value of `nbConnection` always non-empty?

Comment (tomponline) [reply lxd/network/openvswitch/ovn.go]: Ah i see its defined in Config Schema too:

```
"network.ovn.northbound_connection": {Default: "unix:/var/run/ovn/ovnnb_db.sock"},
```

Comment (tomponline) [reply lxd/network/openvswitch/ovn.go]: OK, yes, then I think we need to make the `Default` value empty string, and then update the documentation of that key to explain that when its empty LXD will use a different default value based on the environment:

1. LXD snap will use the MicroOVN env setttings if connected to MicroOVN.
2. Fallback to "unix:/var/run/ovn/ovnnb_db.sock"

Comment (nmezhenskyi) [reply lxd/network/openvswitch/ovn.go]: Updated

File: shared/util.go [hunk 11]
```diff
@@ -258,6 +258,55 @@ func LogPath(path ...string) string {
...
+		// Split into key and value (only on the first "=").
+		key, value, found := strings.Cut(line, "=")
+		if !found {
+			continue
+		}
+
+		// Trim and strip quotes.
+		key = strings.TrimSpace(key)
+		value = strings.TrimSpace(value)
+		value = strings.Trim(value, `"'`)
```
[the remark this answers is not in the document]
Comment (nmezhenskyi) [reply shared/util.go]: Good point, I've extended the `ParseEnvFile()` function to handle more edge cases and added tests for them.

File: shared/util.go [hunk 12]
```diff
@@ -258,6 +258,96 @@ func LogPath(path ...string) string {
...
+	// stripInlineComment removes inline shell comments from a value.
+	// A valid inline comment typically requires a space or tab before the '#' character
+	// to distinguish it from a natural '#' (e.g., in a URL).
+	stripInlineComment := func(val string, prefixToRemove string) string {
+		// Look for a space followed by a '#'.
+		idx := strings.Index(val, " #")
+		if idx == -1 {
+			// If not found, look for a tab followed by a '#'.
+			idx = strings.Index(val, "\t#")
+		}
```
Comment (simondeziel) [inline shared/util.go]: I wasn't sure is a white space was needed but apparently it is:

```
$ FOO=hey#ha; echo $FOO
hey#ha
$ FOO="hey"#ho; echo $FOO
hey#ho
$ FOO='hey'#hu; echo $FOO
hey#hu
```

File: shared/util.go [hunk 13]
```diff
@@ -258,6 +258,96 @@ func LogPath(path ...string) string {
...
+
+			// Find the matching closing quote.
+			endIdx := strings.IndexByte(value[1:], quote)
+
+			if endIdx != -1 {
+				// The value is exactly what is inside the quotes.
+				// Anything after the closing quote (like `# comment`) is naturally ignored.
+				value = value[1 : endIdx+1]
+			} else {
+				// The quote was never closed on this line.
```
Comment (simondeziel) [inline shared/util.go]: The only valid case for unclosed quote is a multiline value which I don't think we want to handle.

I'd outright reject this rather than take a part of it.

Comment (nmezhenskyi) [reply shared/util.go]: Updated. A missing assignment  operator `=` now also returns an error.

**Review by simondeziel — DISMISSED** [review]
LGTM, thanks!
The `ParseEnvFile` helper quickly got more complicated than what initially implemented but I think that's needed due to how flexible shell sourcing is :/

You added good test coverage so that covers our bases nicely.

**Comment by tomponline** [comment]
@nmezhenskyi the env file parsing looks rather complex, please can you paste an example of the file you're trying to parse onto this PR please.

I wonder if there is a simpler way using our snap packaging to extract the env var we want and relay it to LXD as another env var - or perhaps just sourcing the file in the snap packaging and have LXD use the same env var.

Similar to what we do with the gpu-2404 snap plug functionality.

https://github.com/canonical/lxd-pkg-snap/commit/2cfe7da7dfb8ed32ff47558646b774beb002a19c

**Comment by nmezhenskyi** [comment]
> @nmezhenskyi the env file parsing looks rather complex, please can you paste an example of the file you're trying to parse onto this PR please.

@tomponline Here is the example `ovn.env`:

```sh
# # Generated by MicroOVN, DO NOT EDIT.
OVN_INITIAL_NB="10.251.3.75"
OVN_INITIAL_SB="10.251.3.75"
OVN_NB_CONNECT="ssl:10.251.3.75:6641"
OVN_SB_CONNECT="ssl:10.251.3.75:6642"
OVN_LOCAL_IP="10.251.3.75"
```

**Comment by tomponline** [comment]
> > @nmezhenskyi the env file parsing looks rather complex, please can you paste an example of the file you're trying to parse onto this PR please.
> 
> @tomponline Here is the example `ovn.env`:
> 
> ```shell
> # # Generated by MicroOVN, DO NOT EDIT.
> OVN_INITIAL_NB="10.251.3.75"
> OVN_INITIAL_SB="10.251.3.75"
> OVN_NB_CONNECT="ssl:10.251.3.75:6641"
> OVN_SB_CONNECT="ssl:10.251.3.75:6642"
> OVN_LOCAL_IP="10.251.3.75"
> ```

So really we want to start LXD with those env vars exposed to it, and then we can use `OVN_NB_CONNECT` directly right?

**Comment by nmezhenskyi** [comment]
> So really we want to start LXD with those env vars exposed to it, and then we can use `OVN_NB_CONNECT` directly right?

@tomponline Yes, but we also need to allow LXD "monitoring" the changes to these values because `OVN_NB_CONNECT` can change its value.

> I wonder if there is a simpler way using our snap packaging to extract the env var we want and relay it to LXD as another env var - or perhaps just sourcing the file in the snap packaging and have LXD use the same env var.

If I understand this correctly, we would do this only once at snap connection stage? My concern with this approach is that `OVN_NB_CONNECT` value could go stale since MicroOVN is continually updating the `ovn.env` file as the cluster evolves.

**Comment by tomponline** [comment]
> > So really we want to start LXD with those env vars exposed to it, and then we can use `OVN_NB_CONNECT` directly right?
> 
> @tomponline Yes, but we also need to allow LXD "monitoring" the changes to these values because `OVN_NB_CONNECT` can change its value.
> 
> > I wonder if there is a simpler way using our snap packaging to extract the env var we want and relay it to LXD as another env var - or perhaps just sourcing the file in the snap packaging and have LXD use the same env var.
> 
> If I understand this correctly, we would do this only once at snap connection stage? My concern with this approach is that `OVN_NB_CONNECT` value could go stale since MicroOVN is continually updating the `ovn.env` file as the cluster evolves.

Right, good point.

**Comment by tomponline** [comment]
@nmezhenskyi we cannot be the first people to want to parse a bash file for env vars in Go, do we really need to write our own parser?

**Comment by tomponline** [comment]
@nmezhenskyi could we execute bash with that file and a script to print the var out, and then collect the output?

**Comment by nmezhenskyi** [comment]
> @nmezhenskyi we cannot be the first people to want to parse a bash file for env vars in Go, do we really need to write our own parser?

@tomponline True, there is this library [godotenv](https://github.com/joho/godotenv) that does exactly what we need. It's self-contained and does not have further dependencies. 

I guess this is just a trade off between introducing an additional dependency vs maintaining our custom implementation. So we could go with using the library.

**Comment by nmezhenskyi** [comment]
> @nmezhenskyi could we execute bash with that file and a script to print the var out, and then collect the output?

@tomponline Do you mean to do that inside LXD? I think we could

**Comment by simondeziel** [comment]
> @nmezhenskyi could we execute bash with that file and a script to print the var out, and then collect the output?

Using `bash` or `sh` to `source` (`.`) the file means we fully trust the file content to not contain malicious commands, because yeah, sourcing a file will execute directives in it. Even `bash -r` (aka restricted mode) is unsafe.

**Comment by nmezhenskyi** [comment]
@simondeziel Makes sense. I'll proceed with using the [godotenv](https://github.com/joho/godotenv) library then.

**Comment by tomponline** [comment]
> @simondeziel Makes sense. I'll proceed with using the [godotenv](https://github.com/joho/godotenv) library then.

Yes that should be the safest, but can you review the code to check for compatibility with what we need, and for the code quality?

**Comment by simondeziel** [comment]
I'm guilty for some of the additional complexity when I mentioned that `export ` prefix should be ignored.

On second thought, this snap connection is an interface so I think we can expect a stable contract to be honored so maybe we don't need to go overboard and just assume we'll always get the simple and easy version to trivially parse? Worth checking the OVN team to be 100% safe IMHO.

**Comment by tomponline** [comment]
If we can implement a very simple parse for our requirements then there's no need for an external package, but when i look at what we have now, its seems over complicated and trying to be a generic env var parser, at which point we should stop, and think "is there an existing solution for this?".

**Comment by tomponline** [comment]
There's also https://github.com/hashicorp/go-envparse

**Comment by simondeziel** [comment]
> If we can implement a very simple parse for our requirements then there's no need for an external package, but when i look at what we have now, its seems over complicated

Yes, indeed and that's partly my fault and then Copilot also pointed out some shortcomings that led up to this situation.

> and trying to be a generic env var parser, at which point we should stop, and think "is there an existing solution for this?".

Agreed, hence my suggestion to check if that's safe to assume the simple version is what we are guarantied to be provided through that interface.

**Comment by nmezhenskyi** [comment]
@tomponline @simondeziel MicroOVN is indeed using a very simple template for `ovn.env`: https://github.com/canonical/microovn/blob/e860d3abf65c077a1b6524e11459e90e1c10a106/microovn/ovn/environment/environment.go#L26-L32

So, a custom parser specifically for this format is fairly simple.

**Comment by tomponline** [comment]
> @tomponline @simondeziel MicroOVN is indeed using a very simple template for `ovn.env`: https://github.com/canonical/microovn/blob/e860d3abf65c077a1b6524e11459e90e1c10a106/microovn/ovn/environment/environment.go#L26-L32
> 
> So, a custom parser specifically for this format is fairly simple.

https://github.com/hashicorp/go-envparse looks quite nice too

**Comment by nmezhenskyi** [comment]
@tomponline @simondeziel I went with using https://github.com/hashicorp/go-envparse library after all. It is self-contained, small, and has good test coverage.

I've decided against our own simple parser mainly because someone down the road might want to use it for something other than `ovn.env` and will face unexpected problems.

**Review by simondeziel — DISMISSED** [review]
Nice and clean, thanks

**Comment by tomponline** [comment]
@nmezhenskyi we will need an api extension for this so that microcloud can detect and avoid setting northbound connection setting.

**Comment by nmezhenskyi** [comment]
> @nmezhenskyi we will need an api extension for this so that microcloud can detect and avoid setting northbound connection setting.

@tomponline Makes sense, I've added it

File: shared/version/api.go [hunk 14]
```diff
@@ -480,6 +480,7 @@ var APIExtensions = []string{
 	"clustering_control_plane",
 	"storage_remote_drop_source",
 	"storage_ceph_use_rbd_defaults",
+	"ovn_dynamic_northbound_connection",
```
Comment (roosterfish) [inline shared/version/api.go:484]: Looks like we should also to get this into MicroCloud. Meaning if LXD has `ovn_dynamic_northbound_connection`, don't anymore set `network.ovn.northbound_connection` and unset if present.

Comment (roosterfish) [reply shared/version/api.go:484]: Ah, haven't checked the latest comment https://github.com/canonical/lxd/pull/17874#issuecomment-4077210151. nvm.

Comment (roosterfish) [reply shared/version/api.go:484]: Issue for MicroCloud https://github.com/canonical/microcloud/issues/1286.

Comment (nmezhenskyi) [reply shared/version/api.go:484]: Thanks!

Comment (tomponline) [reply shared/version/api.go:484]: thanks @roosterfish indeed, we intended for MicroCloud to be updated to benefit from this. ta

File: lxd/network/openvswitch/ovn.go [hunk 15]
```diff
@@ -180,14 +183,52 @@ type OVNRouterPeering struct {
...
+			defer envFile.Close()
+
+			ovnEnvVars, err := envparse.Parse(envFile)
+			if err != nil {
+				return nil, fmt.Errorf("Failed parsing MicroOVN env file %s: %w", envFilePath, err)
+			}
+
+			nbConn, ok := ovnEnvVars["OVN_NB_CONNECT"]
+			if !ok || nbConn == "" {
+				return nil, errors.New("OVN_NB_CONNECT not found in MicroOVN env file")
```
Comment (tomponline) [inline lxd/network/openvswitch/ovn.go]: Please can you add the file path of the file being parsed here like you did for the other errors?

**Review by tomponline — CHANGES_REQUESTED** [review]
One tiny change in one of the errors, but apart from that this is great, thanks!

File: lxd/network/openvswitch/ovn.go (hunk 15 above)
Comment (nmezhenskyi) [reply lxd/network/openvswitch/ovn.go]: Updated

**Comment by tomponline** [comment]
@nmezhenskyi please rebase

**Review by tomponline — APPROVED** [review]
LGTM thanks!

**Comment by tomponline** [comment]
@nmezhenskyi please can you follow up by updating our test suite to not set `LXD_OVN_NB_CONNECTION` and update `ovn_enabled` and `setup_ovn` to allow for detecting microcloud ovn-env connection and not requiring `LXD_OVN_NB_CONNECTION` be set in those cases.

**Comment by tomponline** [comment]
@roosterfish fyi this is merged now.
